Allow `fields` to be set to `*`

matriv opened the pull request:
Allow for SimpleQueryString, QueryString and MultiMatchQuery
to set the `fields` parameter to the wildcard `*`. If so, set
the leniency to `true`, to achieve the same behaviour as from the
`"default_field" : "*" `setting.

Closes: #39577

matriv committed:
Allow `fields` to be set to `*`

Allow for SimpleQueryString, QueryString and MultiMatchQuery
to set the `fields` parameter to the wildcard `*`. If so, set
the leniency to `true`, to achieve the same behaviour as from the
`"default_field" : "*" setting.

Closes: #39577

elasticmachine commented:
Pinging @elastic/es-search

matriv commented on server/src/main/java/org/elasticsearch/index/query/SimpleQueryStringBuilder.java:
```diff
@@ -399,18 +399,21 @@ public SimpleQueryStringBuilder fuzzyTranspositions(boolean fuzzyTranspositions)
-            if (isAllField) {
-                newSettings.lenient(lenientSet ? settings.lenient() : true);
-            }
+            isAllField = defaultFields.size() == 1 && Regex.isMatchAllPattern(defaultFields.get(0));
```
@jimczi Why do we check only the first element here, shouldn't we check all the `defaultFields`?
(same applies to all 3 types of queries)

jimczi replied:
I agree we should check all elements. However I don't think it's useful to set `fields` or the default fields to something like `["*", "field1", "field2"]` but for consistency we should always set the leniency to true if `*` is part of the list of fields to query.

jimczi commented on server/src/test/java/org/elasticsearch/index/query/MultiMatchQueryBuilderTests.java:
```diff
@@ -419,42 +420,65 @@ public void testDefaultField() throws Exception {
+        }
+        assertEquals(11, noMatchNoDocsQueries);
+        assertThat(disjunctionMaxQuery.getDisjuncts(), hasItems(new TermQuery(new Term(STRING_FIELD_NAME, "hello")),
+            new TermQuery(new Term(STRING_FIELD_NAME_2, "hello"))));
```
Can you also add a test with a single field set on `*` ?

jimczi commented on server/src/test/java/org/elasticsearch/index/query/QueryStringQueryBuilderTests.java:
```diff
@@ -1278,6 +1279,25 @@ public void testDefaultField() throws Exception {
+        }
+        assertEquals(11, noMatchNoDocsQueries);
+        assertThat(disjunctionMaxQuery.getDisjuncts(), hasItems(new TermQuery(new Term(STRING_FIELD_NAME, "hello")),
+            new TermQuery(new Term(STRING_FIELD_NAME_2, "hello"))));
```
Same here

jimczi reviewed (changes requested):
I left some small comments but it looks good @matriv.

matriv committed:
Address comments

Also check if `*` is in the list of the default_field but
not necessarily as the 1st element.

matriv committed:
Merge remote-tracking branch 'upstream/master' into fix-39577

matriv committed:
remove unused imports

cbuescher commented on server/src/test/java/org/elasticsearch/index/query/MultiMatchQueryBuilderTests.java:
```diff
@@ -536,4 +564,18 @@ private static IndexMetaData newIndexMeta(String name, Settings oldIndexSettings
+                noMatchNoDocsQueries++;
+            }
+        }
+        assertEquals(11, noMatchNoDocsQueries);
```
small suggestion: I think we have 11 clauses because that its the number of fields defined in AbstractBuilderTestCase#MAPPED_LEAF_FIELD_NAMES. To make the test more robust to future changes, maybe use that array size here?

matriv replied:
I thought about it but I think that's not correct, both `MAPPED_LEAF_FIELD_NAMES` and `MAPPED_FIELD_NAMES` contain less fields + some aliases and are not used for the registration of the fields. The registration takes place here: https://github.com/elastic/elasticsearch/blob/master/test/framework/src/main/java/org/elasticsearch/test/AbstractBuilderTestCase.java#L388 and doesn't use either of those arrays.

cbuescher replied:
Bummer. And I cannot see another simple way to get to this value from the shard context eithter, other than using the same code that is used by the query builder, and that would defeat the purpose. Guess it needs to stay like that then.

cbuescher commented on server/src/test/java/org/elasticsearch/index/query/QueryStringQueryBuilderTests.java:
```diff
@@ -1513,4 +1544,18 @@ private static IndexMetaData newIndexMeta(String name, Settings oldIndexSettings
+                noMatchNoDocsQueries++;
+            }
+        }
+        assertEquals(11, noMatchNoDocsQueries);
```
same here.

cbuescher commented on server/src/test/java/org/elasticsearch/index/query/SimpleQueryStringBuilderTests.java:
```diff
@@ -739,4 +772,18 @@ private static IndexMetaData newIndexMeta(String name, Settings oldIndexSettings
+                noMatchNoDocsQueries++;
+            }
+        }
+        assertEquals(11, noMatchNoDocsQueries);
```
and same here.

cbuescher reviewed (changes requested):
The main part looks good to me, I left a very small suggestion about testing.

cbuescher reviewed (approved):
LGTM

jimczi reviewed (approved):
LGTM too
